### PR TITLE
[sw] DRAFT: First pass at a ROM_EXT parsing library

### PR DESCRIPTION
**Pull request by mcy**
This is the very first pass at what the ROM_EXT executable format parser might look like. I still need to transcribe the format spec into the header. Right now, the goals for this library are:
- Writing down an interface.
- Making sure that interface does not depend on anything OpenTitan-ey, so that it can be fuzzed on amd64.
- Making the interface agnostic to the cryptography being used.
- Making sure that glitch resistance can be easily added to the parser and the interface with minimal work.

### CODE REVIEW

File: sw/device/lib/runtime/check.h [hunk 1]
```diff
@@ -37,6 +37,6 @@
  * @param ... arguments to a LOG_* macro, which are evaluated if the check
  * fails.
  */
-#define CHECKZ(value, ...) CHECK((value) == 0, ...)
```
Comment (gkelly) [inline sw/device/lib/runtime/check.h]: Should be a separate commit.

File: sw/device/boot_rom2/ext_format.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  CHECK(crypto.hash_data(crypto.self, metadata, metadata_len),
+        "Failed to hash metadata.");
+  CHECK(crypto.hash_data(crypto.self, payload, header.payload_len),
+        "Failed to hash payload.");
+  CHECK(crypto.hash_finish(crypto.self, &footer_hash_computed),
+        "Failed to complete footer hash.");
+
+  // NOTE: This comparison should probably be made constant-time, since we might
+  // not want an attacker to learn how close a failed hash is.
+  CHECKZ(memcmp(footer_hash, footer_hash_computed, crypto.hash_len),
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: What is the benefit of adding an additional comparison instead of just applying the verify op on the overall hash?  This component is monolithic  anyway.

File: sw/device/boot_rom2/ext_format.c [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+  // Compute offset pointers into |binary| pointing to relevant fields within.
+  const uint8_t *signature = binary + sizeof(header);
+  const uint8_t *footer_hash = signature + crypto.sig_len;
+  const uint8_t *metadata = footer_hash + crypto.hash_len;
+  size_t metadata_len = metadata_len_from_version(version);
+  const uint8_t *payload = metadata + metadata_len;
+
+  // Verify the magic word.
+  CHECK(header.magic == kExtMagic, "Magic value mismatch.");
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: Do we also LOG in production mode? We may have to consider having an option to disable logging in ROM/ROM_EXT based on fuse configuration, or have a debug logging mode.

File: sw/device/boot_rom2/ext_format.c [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+   * Absolute address of the start of the payload region.
+   */
+  const uint8_t *payload;
+  /**
+   * Length of the payload region.
+   */
+  size_t payload_len;
+  /**
+   * Absolute address of the next stage entrypoint.
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: nit: Is this the entry point of the code partition being verified?  Why not say something more explicit like "payload entrypoint"?

File: sw/device/boot_rom2/ext_format.h [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+} ext_crypto_vt_t;
+
+/**
+ * Metadata value for the ROM_EXT-v0 format.
+ */
+typedef struct ext_metadata_v0 {
+  /**
+   * The entrypoint to the next stage of boot.
+   */
+  void *entrypoint;
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.h]: What is the difference between this and `header_t.entrypoint`?

File: sw/device/boot_rom2/ext_format.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
+// Cn:opyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/boot_rom2/ext_format.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/runtime/check.h"
+
+static const uint32_t kExtMagic = 0x5245545f;
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: nit: ascii = "RET_" ? Where did this come from? Comment if it makes sense to add context.

File: sw/device/boot_rom2/ext_format.c [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ * @param crypto a cryptographic vtable.
+ * @param version the version this parse is for (which is used to determine the
+ * metadata length).
+ * @return "common" parsed out values.
+ */
+static common_parse_result_t common_parse(const uint8_t *binary,
+                                          ext_crypto_vt_t crypto,
+                                          ext_format_version_t version) {
+  fixed_header_t header;
+  memcpy(&header, binary, sizeof(header));
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: What is the reasoning behind copying the header?

File: sw/device/boot_rom2/ext_format.c [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/runtime/check.h"
+
+static const uint32_t kExtMagic = 0x5245545f;
+
+/**
+ * Convenience struct for pulling out the fixed part of the executable header.
+ */
+typedef struct fixed_header {
+  uint32_t magic, version, payload_len, entrypoint;
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: There are other things that will be added here later, but a key_id or hash of the pub_key used to verify the signature may be useful to integrate since rom will have multiple keys (potentially with different purposes, e.g. TEST versus DEV versus PROD ).

File: sw/device/boot_rom2/ext_format.c [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+ * @return "common" parsed out values.
+ */
+static common_parse_result_t common_parse(const uint8_t *binary,
+                                          ext_crypto_vt_t crypto,
+                                          ext_format_version_t version) {
+  fixed_header_t header;
+  memcpy(&header, binary, sizeof(header));
+
+  // Compute offset pointers into |binary| pointing to relevant fields within.
+  const uint8_t *signature = binary + sizeof(header);
```
Comment (moidx) [inline sw/device/boot_rom2/ext_format.c]: nit: It would be nice to document the expected layout of the binary somewhere... I guess this will eventually point to a linker script or image header...

File: sw/device/boot_rom2/ext_format.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: Wouldn't some other code be responsible for deciding which of A or B is active? I don't think the actual executable needs to know whether it is an A or a B?

File: sw/device/boot_rom2/ext_format.c (hunk 2 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: I'm not sure I follow. Here's the scheme I'm using right now:
- The fixed-location components are authenticated by a signature, which can be verified without looking at any other values in the binary (besides reading them to put them into the HMAC engine, of course).
- The variable-length parts (the metadata + the payload) have a hash that can be computed using only values directly protected by the first signature. The hash itself is protected by the signature.

There isn't an overall signature. I've pushed a change that adds a high-level description of the format and its security model, which I think might give more context than a bunch of C code.

File: sw/device/boot_rom2/ext_format.c (hunk 3 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: Currently, yes, but that's because there really isn't a "production mode". Eventually, it will work how you describe (though I think that's a bit out of scope; the CHECKs are, more than anything, for illustration purposes right now).

File: sw/device/boot_rom2/ext_format.c (hunk 4 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: Yes. I thought that that would strongly implicit, given the parallel with loading an ELF file? I can change it if you still disagree.

File: sw/device/boot_rom2/ext_format.h (hunk 5 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.h]: header_t.entrypoint is supposed to be an offset from the start of the payload (so that the payload doesn't need to know where it is being loaded at). Perhaps that should be renamed to entrypoint_offset?

File: sw/device/boot_rom2/ext_format.c (hunk 6 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: It's little endian, so the ascii is actually "_TER". THat's actually a typo on my part, too: the first character should be an O, for "OpenTitan Ext Rom".

Good catch.

File: sw/device/boot_rom2/ext_format.c (hunk 7 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: This is the only legal way to actually read it, due to alignment constraints[1]. fixed_header_t just exists to make this copy shorter in code.

[1] In C, fixed_header_t has an alignment of 4, but in the format, an alignment of 1 (since the format doesn't believe in alignment).

File: sw/device/boot_rom2/ext_format.c (hunk 8 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: This is a first pass of the format. I'm expecting that kind of stuff (and whether it should go in the permanent header or in versioned metadata) to fall out of a more thorough security review.

File: sw/device/boot_rom2/ext_format.c (hunk 9 above)
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: I've added some documentation about the format and the encoding in the header. Hopefully that's enough context for now (I have a doc with even more context that I can share).

File: sw/device/lib/runtime/check.h (hunk 1 above)
Comment (mcy) [reply sw/device/lib/runtime/check.h]: It's going to be a separate PR. I just haven't gotten around to it, because this is still a draft.

File: sw/device/boot_rom2/ext_format.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (moidx) [reply sw/device/boot_rom2/ext_format.c]: It depends on how error handling is implemented. Usually, if partition A fails to boot, the firmware may attempt to boot partition B if it is qualified to run. This has implications on how firmware update is implemented as well.  

Also note that we are not planning to have a recovery partition at this moment, so we need to handle issues with the active partition gracefully.

File: sw/device/boot_rom2/ext_format.c (hunk 8 above)
Comment (moidx) [reply sw/device/boot_rom2/ext_format.c]: ack. Re. the versioned metadata comment, the key identifier should go with the signature.

File: sw/device/boot_rom2/ext_format.c (hunk 2 above)
Comment (moidx) [reply sw/device/boot_rom2/ext_format.c]: OK, just for practical purposes assume that the flash page size is 4KB.  The ROM_EXT will be aligned to page size granularity to simplify handling of flash erase during firmware update etc. Thus in practice, the software image hash usually includes even the padding at the end of the image to the page size boundary.

The signature is also aligned to a 32bit multiple.  

Using this constraints, you can reduce the number of memory comparisons and simplify hardening. Usually what is done with monolithic components like this one is to aggregate all the hashes (SW, HW measurements, Other Measurements) using a cryptographic function, and  feed them into the verify function. 

The aggregated hash can then be used in a warm boot implementation that relies on a lighter verification scheme.

File: sw/device/boot_rom2/ext_format.h (hunk 5 above)
Comment (moidx) [reply sw/device/boot_rom2/ext_format.h]: I am marking this as a nit. You can rename entrypoint_offset in header_t if you consider that it improves readability.

File: sw/device/boot_rom2/ext_format.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: I mean, what I would like is that all parse errors are treated as some flavor of fatal error that kicks control all the way back to the start of boot, perhaps with some kind of signaling.

I might be ok with some scheme where *any* error is signaled as a return value of "something went wrong, it's already been logged so all you get to know is that *something* happened". I might just replace the CHECKs with some kind of similar macro that handles the unhappy path... I just want error handling to be "really boring".

[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/boot_rom2/ext_format.c]: I've gone through and made it so that the parsing routines produce an opaque error on failure, which is definitely good enough for A/B updates.

Also, now we get to promise "never crashes" as a constraint for fuzzing.

**Comment by lenary** [comment]
We've dropped work on `boot_rom2`, so I'm closing this.
